File: extensions/youtube/src/main/java/app/argusos/extension/youtube/patches/DisablePlayerPopupPanelsPatch.java
package app.argusos.extension.youtube.patches;

import app.morphe.extension.shared.Logger;
import app.argusos.extension.youtube.settings.Settings;

@SuppressWarnings("unused")
public class DisablePlayerPopupPanelsPatch {

    /**
     * Injection point.
     */
    public static boolean disablePlayerPopupPanels() {
        if (Settings.DISABLE_PLAYER_POPUP_PANELS.get()) {
            Logger.printDebug(() -> "disablePlayerPopupPanels: Popup panels blocked!");
            return true;
        }
        return false;
    }
}

File: patches/src/main/kotlin/app/argusos/patches/youtube/layout/hide/player/popup/DisablePlayerPopupPanelsPatch.kt
package app.argusos.patches.youtube.layout.hide.player.popup

import app.morphe.patcher.extensions.InstructionExtensions.addInstructionsWithLabels
import app.morphe.patcher.patch.bytecodePatch
import app.argusos.patches.shared.misc.settings.preference.SwitchPreference
import app.argusos.patches.youtube.misc.extension.sharedExtensionPatch
import app.argusos.patches.youtube.misc.settings.PreferenceScreen
import app.argusos.patches.youtube.misc.settings.settingsPatch
import app.argusos.patches.youtube.shared.Constants.COMPATIBILITY_YOUTUBE

private const val EXTENSION_CLASS =
    "Lapp/argusos/extension/youtube/patches/DisablePlayerPopupPanelsPatch;"

@Suppress("unused")
val disablePlayerPopupPanelsPatch = bytecodePatch(
    name = "Disable player popup panels",
    description = "Adds an option to disable panels (such as live chat) from opening automatically.",
) {
    dependsOn(
        sharedExtensionPatch,
        settingsPatch,
    )

    compatibleWith(COMPATIBILITY_YOUTUBE)

    execute {
        PreferenceScreen.PLAYER.addPreferences(
            SwitchPreference("argusos_disable_player_popup_panels", summary = true),
        )

        PlayerPopupPanelsFingerprint.method.addInstructionsWithLabels(
            0,
            """
                invoke-static { }, $EXTENSION_CLASS->disablePlayerPopupPanels()Z
                move-result v0
                if-eqz v0, :player_popup_panels
                return-void
                :player_popup_panels
                nop
            """
        )
    }
}
